src/block_cache/block_cache.c

```diff
@@ -505,21 +505,32 @@ __wt_blkcache_put(
  * __wt_blkcache_remove --
  *     Remove a block from the cache.
  */
-void
+int
 __wt_blkcache_remove(WT_SESSION_IMPL *session, const uint8_t *addr, size_t addr_size)
 {
     WT_BLKCACHE *blkcache;
     WT_BLKCACHE_ITEM *blkcache_item;
-    uint64_t bucket, hash;
+    uint64_t bucket, hash, sleep_usecs, yield_count;
 
     blkcache = &S2C(session)->blkcache;
-
     hash = __wt_hash_city64(addr, addr_size);
     bucket = hash % blkcache->hash_size;
+    sleep_usecs = yield_count = 0;
+
     __wt_spin_lock(session, &blkcache->hash_locks[bucket]);
     TAILQ_FOREACH (blkcache_item, &blkcache->hash[bucket], hashq) {
         if (blkcache_item->addr_size == addr_size && blkcache_item->fid == S2BT(session)->id &&
           memcmp(blkcache_item->addr, addr, addr_size) == 0) {
+            /*
+             * The block might be in use by another thread, wait for it to be released before
+             * freeing it.
+             */
+            while (blkcache_item->ref_count != 0) {
```

Comment:
Is this at the right spot? The assertions and the `__blkcache_free` calls are a few lines below and I am not sure if I should bring this `while` loop closer.
I added this loop here as we are removing the block using `TAILQ_REMOVE` at the beginning of this section and we should not do that while a block is in use.

Reply:
I'm definitely not an expert in this area so if anyone knows better feel free to push back. I'm looking at where you've put the spin lock and agree it should be before the `TAILQ_REMOVE`, but I'm wondering if it's possible for us to shift all the statistic updates so that they are after the `blkcache_item->ref_count == 0` assert and the call to `__blkcache_free ` if it is not as important for these statistics to be updated at that point in time. Although I think the window is already quite short as is, this would shorten the window even further to prevent any other races.

Reply:
I agree, see here 4d12c7313e601e312b2065089ffa1bb80099ecb2.